### PR TITLE
[Feature] Support batch job aware scheduler

### PR DESCRIPTION
**Pull request by wanqi01**
### What this PR does / why we need it?
This PR introduces a **Batch-Job-Aware Scheduler** designed for **offline batch inference** scenarios where throughput and hardware utilisation are the primary goals. It is particularly effective when processing multiple batch jobs concurrently, each with a distinct set of requests.

#### Background

In large-scale offline batch processing scenarios, the default FCFS (First-Come-First-Serve) scheduler suffers from the following issues:

1. **Low throughput during decode-heavy phases**: When a large number of requests in the running queue are in the decoding phase, the number of tokens computed per scheduling round is small, resulting in low throughput.

2. **Excessive preemption overhead**: When running queue requests are scheduled, if a request cannot acquire new KV cache blocks, preemption is triggered, which can cause a large number of requests to be recomputed, wasting computational resources.

Since the request lists of each job type share the same prompt template and task objective, their decode-phase token lengths tend to converge. We can predict a job's decode-phase token length by sampling its completed requests, and incorporate these predictions into scheduling optimization.

#### Key Design

The scheduler implements three key strategies to improve throughput:

1. **LPT (Longest Processing Time first) scheduling**: Prioritises longer tasks first and schedules shorter ones to fill gaps, particularly during the decode step. This improves the average number of tokens computed per scheduling round.

2. **KV cache reservation**: Estimates and reserves KV cache budget in advance for running requests. The `RunningBlockReserver` predicts the next-step KV block needs for each running request, using a two-level incremental cache (O(1) for single-request additions within the same admission loop) to avoid redundant full scans. This reduces preemption overhead significantly.

3. **Job-aware request grouping with dynamic priority**: Groups requests by **job name** (extracted from a `#job_name[${JOB_NAME}]#` tag embedded in the request ID), assigns each job its own `RequestBucket` for O(log n) best-fit prefill-length search, and dynamically adjusts job scheduling order based on KV cache availability:
   - When available tokens > threshold (default 4096): prioritise long decode jobs
   - When available tokens ≤ threshold: prioritise short decode jobs

> **Note:** This scheduler **does not implement starvation prevention**. It is intended solely for **batch processing scenarios** or **online inference scenarios where request waiting time is not a concern**. If requests have strict latency or fairness requirements, this scheduler may not be suitable.

#### Architecture

The scheduler is implemented by extending the **waiting queue** with a custom `BatchJobAwareRequestQueue` that replaces the default FIFO queue, while inheriting most features from the original scheduler, including chunked prefill, async scheduling, and more.

Key components:

| Component | Description |
|-----------|-------------|
| `JobNameParser` | Extracts job names from request IDs with regex caching |
| `JobDecodeEstimator` | Per-job decode length prediction using EWMA + Bayesian cold-start shrinkage |
| `RunningBlockReserver` | KV cache demand prediction with incremental cache updates |
| `RequestBucket` | O(log n) best-fit search by prefill length |
| `BatchJobAwareRequestQueue` | Job-grouped waiting queue with reserve-aware admission |
| `BatchJobAwareScheduler` | Main scheduler class (sync) |
| `BatchJobAwareAsyncScheduler` | Combined async + batch-job-aware scheduler |

#### Decode Length Estimation

Each job's decode length is predicted using a pure **EWMA (Exponentially Weighted Moving Average)** estimator:

- **No samples yet**: Returns a cold-start default value (128 tokens).
- **First observation**: EWMA is initialised to the observed decode length.
- **Subsequent observations**: EWMA is updated incrementally, giving more weight to recent data while smoothing out fluctuations.

This enables the scheduler to distinguish between "long decode" jobs (which benefit from being scheduled first when resources are abundant) and "short decode" jobs (which are prioritised when resources are scarce).

#### Configuration

All parameters are nested under `batch_job_sched_config` in the `additional_config` dictionary:

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `enabled` | bool | `false` | Enable the batch-job-aware scheduler |
| `max_jobs` | int | `20` | Maximum number of tracked jobs |
| `reserve_margin_blocks` | int | `2` | Extra block margin added to the KV cache reserve as safety buffer |
| `reserve_max_blocks` | int | `8` | Maximum number of blocks that can be reserved |
| `low_available_tokens_threshold` | int | `4096` | Threshold for prioritising long vs short decode jobs |
| `short_decode_token_threshold` | int | `32` | Threshold for classifying a job as "short decode" |

#### Files Changed

| File | Change |
|------|--------|
| `vllm_ascend/core/batch_job_aware_scheduler.py` | **New**: Core scheduler implementation  |
| `vllm_ascend/ascend_config.py` | **Modified**: Added `BatchJobSchedConfig` class and config loading |
| `vllm_ascend/platform.py` | **Modified**: Register scheduler classes when config enabled  |
| `docs/source/user_guide/feature_guide/batch_job_aware_scheduler.md` | **New**: Feature usage documentation  |
| `docs/source/user_guide/configuration/additional_config.md` | **Modified**: Add config documentation  |
| `tests/ut/core/test_batch_job_aware_scheduler.py` | **New**: Unit tests  |
| `tests/e2e/pull_request/one_card/test_batch_job_aware_scheduler_e2e.py` | **New**: End-to-end tests |
| `‎mkdocs.yml` | **Modified**: Add doc index for user_guide |

**Total**: 8 files changed.

---

### Does this PR introduce _any_ user-facing change?

Yes.

- **New configuration**: Adds `batch_job_sched_config` section to `additional_config` with 6 configurable parameters.
- **New request ID convention**: Users can optionally embed `#job_name[${JOB_NAME}]#` in request IDs to enable job-aware grouping.
- **New feature guide**: Added `docs/source/user_guide/feature_guide/batch_job_aware_scheduler.md` with usage documentation.
- **New feature index entry**: Added `batch_job_aware_scheduler` to the feature guide toctree.

Feature is disabled by default (`enabled: false`). Users opt in via `additional_config`:

```bash
python -m vllm.entrypoints.openai.run_batch \
    --model /path/to/model \
    -i /path/to/input.jsonl \
    -o /path/to/output.jsonl \
    --additional-config '{"batch_job_sched_config": {"enabled": true}}'
```

Or via Python API:

```python
from vllm import LLM

llm = LLM(
    model="/path/to/model",
    additional_config={
        "batch_job_sched_config": {
            "enabled": True,
        },
    },
)
```

---

#### Unit Tests

Added comprehensive unit tests in `tests/ut/core/test_batch_job_aware_scheduler.py` covering all components:

- **`_cdiv`** (6 tests): Exact division, rounding up, zero numerator, large values.
- **`_JobStat`** (2 tests): Initial values, after observations.
- **`JobNameParser`** (7 tests): With/without job name, caching, removal, clear, multiple jobs.
- **`JobDecodeEstimator`** (12 tests): Default prediction, observation, cold-start Bayesian shrinkage, stable EWMA phase, EWMA convergence, max_jobs limit, max_samples_per_job limit, get_stats, multiple independent jobs, cumulative sum.
- **`RequestBucket`** (16 tests): Put, remove, best-fit search, empty state, iteration, bool conversion, edge cases.
- **`RunningBlockReserver`** (8 tests): Cache invalidation, empty running, cache hit, incremental update, full recompute, prefill/decode phase computation, block boundary, max blocks cap.
- **`BatchJobAwareRequestQueue`** (22 tests): Add/remove request, job routing, cold-start tracking, peek/pop, job classification, sorting strategies, admission budget, bool conversion, max_jobs limit.

All tests can be run with:

```bash
# Unit tests
pytest -sv tests/ut/core/test_batch_job_aware_scheduler.py
```

#### End-to-End Tests

Added E2E tests in `tests/e2e/pull_request/one_card/test_batch_job_aware_scheduler_e2e.py` with 4 test scenarios:

1. **Basic functionality**: Compares outputs between `BatchJobAwareScheduler` and default scheduler to ensure correctness (4 prompts).
2. **Async scheduling**: Tests `BatchJobAwareAsyncScheduler` against default async scheduler.
3. **Chunked prefill**: Verifies correctness with `enable_chunked_prefill=True` and `chunked_prefill_token_size=16` (20 prompts).
4. **Custom configuration**: Tests with custom EWMA parameters and other config options (20 prompts).

All E2E tests compare outputs using `check_outputs_equal` to verify that the batch-job-aware scheduler produces identical results to the default scheduler, ensuring correctness.

E2E tests can be run with:

```bash
# End-to-end tests (requires Ascend NPU)
pytest -sv tests/e2e/pull_request/one_card/test_batch_job_aware_scheduler_e2e.py
```

#### Performance Tests

We conducted benchmark tests comparing the **Batch-Job-Aware Scheduler** (experimental group) against the **default FCFS scheduler** (control group) in offline batch inference scenarios. Each test group was run 3 times and the average metrics are reported.

**Environment:**
- Framework: vLLM v0.21.0, vLLM-Ascend v0.21.0rc1
- Hardware: Ascend 910B3 NPU (single card)
- Model: Qwen3-14B (deep thinking disabled)
- GPU memory utilization: 0.8
- Chunked prefill enabled, max_num_seqs=500, max_num_batched_tokens=4096

**Test workloads:**

| Workload | Description | Prefill length | Avg decode length |
|----------|-------------|---------------|-------------------|
| News summarization (10,000 requests) | Summarise news in ~100 words | 500–10,000+ | 96 |
| Sentiment classification (30,000 requests) | Classify product review sentiment (positive/negative) | 40–100 | 3 |

- News summarization data: [Chinese Financial News 2019](https://www.modelscope.cn/datasets/tangzhiling/chinese-financial-news-2019/files)
- Sentiment classification data: [JD Product Reviews](https://www.modelscope.cn/datasets/DAMO_NLP/jd/file/view/master/dev.csv?id=350&status=1)

**Results:**

| Workload | Async | Metric | Control (FCFS) | Experimental (Batch-Job-Aware) | Improvement |
|----------|-------|--------|----------------|-------------------------------|-------------|
| News (3K) + Sentiment (30K) | Yes | Avg tokens/schedule | 2,107 | 2,587 | **+22.78%** |
| | | Preemption count | 65 | 1 | **−98.46%** |
| | | Total time (s) | 934.85 | 901.77 | +3.54% |
| News (3K) + Sentiment (30K) | No | Avg tokens/schedule | 2,090 | 2,634 | **+26.03%** |
| | | Preemption count | 73 | 1 | **−98.63%** |
| | | Total time (s) | 956.85 | 923.01 | +3.53% |
| News (10K) | Yes | Avg tokens/schedule | 1,471 | 1,455 | −1.09% |
| | | Preemption count | 220 | 23 | **−89.55%** |
| | | Total time (s) | 1,697.77 | 1,647.67 | +2.95% |
| News (10K) | No | Avg tokens/schedule | 1,468 | 1,465 | −0.20% |
| | | Preemption count | 190 | 19 | **−90.00%** |
| | | Total time (s) | 1,748.18 | 1,699.59 | +2.78% |

> **Note:** For the News (10K) workload, `reserve_margin_blocks` was set to 4 to accommodate longer decode sequences.

**Key findings:**
1. **Significant reduction in preemption**: The batch-job-aware scheduler reduces preemption counts by 89–98%, even in single-job workloads (News 10K) where preemption is reduced by ~90%, thanks to the KV cache reservation mechanism.
2. **Higher tokens per scheduling round in mixed workloads**: Average tokens computed per scheduling round increases by 22–26% in mixed workloads, as the LPT scheduling and job-aware grouping enable better batching of decode-phase requests. In single-job workloads (News 10K), the avg tokens/schedule remains comparable to the baseline while preemption is dramatically reduced.
3. **Consistent throughput improvement**: Overall end-to-end time is reduced by 2.8–3.5% across all workloads. Notably, in the News (10K) workload, the expanded test coverage (both async and non-async modes) shows consistent ~2.8–3.0% time reduction, achieved primarily through eliminating preemption overhead rather than increasing per-round token count.

The results confirm that the batch-job-aware scheduler is particularly effective in multi-job mixed workloads (e.g., news summarisation + sentiment classification) where job decode lengths vary significantly, achieving both higher throughput and dramatically fewer preemptions. It also delivers consistent gains in single-job workloads by minimising preemption overhead.

- vLLM version: v0.24.0
- vLLM main: https://github.com/vllm-project/vllm/commit/85c09e9885e346ea1612da30ebff5a75f67d2350

### CODE REVIEW

File: vllm_ascend/ascend_config.py [hunk 1]
```diff
@@ -721,6 +725,110 @@ def _validate(self):
             raise ValueError(f"Recommend to use at least 30 data points for fitting, got {self.max_fit_chunk}")
 
 
+class BatchJobSchedConfig:
```
Comment (zzzzwwjj) [inline vllm_ascend/ascend_config.py:692]: Please add params description for BatchJobSchedConfig in `additional_config.md`.

File: vllm_ascend/ascend_config.py [hunk 2]
```diff
@@ -721,6 +725,110 @@ def _validate(self):
...
+        self.reserve_margin_blocks: int = int(config.get("reserve_margin_blocks", 2))
+
+        # Maximum number of blocks that can be reserved
+        self.reserve_max_blocks: int = int(config.get("reserve_max_blocks", 8))
+
+        # Threshold for prioritizing long vs short decode jobs
+        self.low_available_tokens_threshold: int = int(config.get("low_available_tokens_threshold", 4096))
+
+        # Threshold for identifying short decoding jobs
+        self.short_decode_token_threshold: int = int(config.get("short_decode_token_threshold", 32))
```
Comment (zzzzwwjj) [inline vllm_ascend/ascend_config.py:723]: There are too many params in BatchJobSchedConfig, please reduce the number of param.

File: docs/source/user_guide/feature_guide/batch_job_aware_scheduler.md
Comment (zzzzwwjj) [inline docs/source/user_guide/feature_guide/batch_job_aware_scheduler.md:1]: Since this scheduler lacks starvation prevention handling, it is advisable to state in the documentation that this scheduling is only suitable for batch processing scenarios or online inference scenarios where the waiting time for requests is not of concern.

**Review by zzzzwwjj — CHANGES_REQUESTED** [review]

**Review by zzzzwwjj — APPROVED** [review]

**Comment by wanqi01** [comment]
The branch is not clean,resubmit the PR
